sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -81,156 +69,103 @@
+    if (matches.size() == 0) {
+      if (required) {
+        throw new IllegalArgumentException(
+            String.format(
+                "No method annotated with @%s found in %s",
+                anno.getSimpleName(), fnClazz.getName()));
+      } else {
+        return null;
       }
     }
-  }
 
-  private static final Map<Class<?>, AdditionalParameter> EXTRA_CONTEXTS;
-  private static final Map<Class<?>, AdditionalParameter> EXTRA_PROCESS_CONTEXTS;
-
-  static {
-    ImmutableMap.Builder<Class<?>, AdditionalParameter> everywhereBuilder =
-        ImmutableMap.<Class<?>, AdditionalParameter>builder();
-    ImmutableMap.Builder<Class<?>, AdditionalParameter> processElementBuilder =
-        ImmutableMap.<Class<?>, AdditionalParameter>builder();
-
-    for (AdditionalParameter value : AdditionalParameter.values()) {
-      switch (value.availability) {
-        case EVERYWHERE:
-          everywhereBuilder.put(value.rawType, value);
-          break;
-        case PROCESS_ELEMENT_ONLY:
-          processElementBuilder.put(value.rawType, value);
-          break;
+    // If we have at least one match, then either it should be the only match
+    // or it should be an extension of the other matches (which came from parent
+    // classes).
+    Method first = matches.iterator().next();
+    for (Method other : matches) {
+      if (!first.getName().equals(other.getName())
+          || !Arrays.equals(first.getParameterTypes(), other.getParameterTypes())) {
+        throw new IllegalArgumentException(
+            String.format(
+                "Found multiple methods annotated with @%s. [%s] and [%s]",
+                anno.getSimpleName(), format(first), format(other)));
       }
     }
 
-    EXTRA_CONTEXTS = everywhereBuilder.build();
-    EXTRA_PROCESS_CONTEXTS = processElementBuilder
-        // Process Element contexts include everything available everywhere
-        .putAll(EXTRA_CONTEXTS)
-        .build();
+    // We need to be able to call it. We require it is public.
+    if ((first.getModifiers() & Modifier.PUBLIC) == 0) {
+      throw new IllegalArgumentException(format(first) + " must be public");
+    }
+
+    // And make sure its not static.
+    if ((first.getModifiers() & Modifier.STATIC) != 0) {
+      throw new IllegalArgumentException(format(first) + " must not be static");
+    }
+
+    return first;
   }
 
-  /**
-   * @return true if the reflected {@link DoFn} uses a Single Window.
-   */
+  /** @return true if the reflected {@link DoFn} uses a Single Window. */
   public abstract boolean usesSingleWindow();
 
   /** Create an {@link DoFnInvoker} bound to the given {@link OldDoFn}. */
   public abstract <InputT, OutputT> DoFnInvoker<InputT, OutputT> bindInvoker(
       DoFn<InputT, OutputT> fn);
 
-  private static final Map<Class<?>, DoFnReflector> REFLECTOR_CACHE =
-      new LinkedHashMap<Class<?>, DoFnReflector>();
+  private static final Map<Class<?>, DoFnReflector> REFLECTOR_CACHE = new LinkedHashMap<>();
```

Comment:
Reflector instances previously filled the purpose of your DoFnSignature. We cached them because we didn't want to do the reflection multiple times. If we're splitting things, I think this should be a Map<Class<?>, DoFnSignature>, that just caches the signature we produced from a given DoFn.